aiida/engine/processes/functions.py

```diff
@@ -228,9 +228,29 @@ def _define(cls, spec):
                 '_node_class': node_class
             })
 
+    @classmethod
+    def validate_inputs(cls, *args, **kwargs):  # pylint: disable=unused-argument
+        """
+        Validate the positional and keyword arguments passed in the function call.
+
+        :raises ValueError: if the passed arguments are invalid
+        """
+        nargs = len(args)
+        nparameters = len(cls._func_args)
+
+        # If the spec is dynamic, i.e. the function signature includes `**kwargs` and the number of positional arguments
+        # passed is larger than the number of explicitly defined parameters in the signature, the inputs are invalid and
+        # we should raise. If we don't, some of the passed arguments, intended to be positional arguments, will be
+        # misinterpreted as keyword arguments, but they won't have an explicit name to use for the link label, causing
+        # the input link to be completely lost.
+        if cls.spec().inputs.dynamic and nargs > nparameters:
+            raise ValueError('function has {} positional arguments, but received only {}'.format(nparameters, nargs))
```

Comment:
I would try to return the same exception that Python returns:
```
In [1]: def f(a,b,**kwargs):
   ...:     pass
   ...: 

In [2]: f(1,2,3)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-b61741968103> in <module>()
----> 1 f(1,2,3)

TypeError: f() takes exactly 2 arguments (3 given)
```

Reply:
Fair enough, bit weird that they chose `TypeError` for this though. Is it because the "type" of `3` is positional whereas it should be keyword? Not sure what the reasoning was

Reply:
I always had the same question in mind :-) Maybe just double check in python3 should they have changed something (I think I tested on py2 above)

Reply:
Same exception but they did improve the message:
```
In [1]: def test(a, **kwargs): 
   ...:     pass 
   ...:                                                                                                                                       

In [2]: test(1, 2)                                                                                                                            
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-f4ef31738bb7> in <module>
----> 1 test(1, 2)

TypeError: test() takes 1 positional argument but 2 were given
```

Reply:
Maybe I should change my message to be the same, I like it, it is clear